File: ProjectSimda/simda/src/main/java/ssafy/a709/simda/dto/FeedDto.java
package ssafy.a709.simda.dto;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import ssafy.a709.simda.domain.Comment;
import ssafy.a709.simda.domain.Feed;
import ssafy.a709.simda.domain.User;

import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.TimeZone;


@Builder
@AllArgsConstructor
@NoArgsConstructor
@Data
public class FeedDto {
    // Feed Id
    private int feedId;
    // Title
    private String title;
    // Emotion
    private int emotion;
    // Content
    private String content;
    // Image Address
    private String img;
    // Latitude
    private double lat;
    // Longitude
    private double lng;
    // Linke Point
    private int likeCnt;
    // Regist Date
    private String regDate;
    //Writer Nick
    private String nickname;
    //Writer id
    private int userId;
    // Comment
    private List<CommentDto> comments;
    // Feed(Entity)를 FeedDto로 변환
    public static FeedDto changeToFeedDto(Feed feed, User user, List<CommentDto> comments){
        if (comments == null) {
            comments = new ArrayList<>(); // 댓글이 없는 경우 빈 리스트 생성
        }

        return FeedDto.builder()
                .feedId(feed.getFeedId())
                .title(feed.getTitle())
                .emotion(feed.getEmotion())
                .nickname(user==null?"탈퇴한 유저입니다.":user.getNickname())
                .userId(user==null?null:user.getUserId())
                .content(feed.getContent())
                .img(feed.getImg())
                .lat(feed.getLat())
                .lng(feed.getLng())
                .likeCnt(feed.getLikeCnt())
                .regDate(convertUtcToSeoul(feed.getRegDate()))
                .comments(comments)
                .build();
    }

    private static String convertUtcToSeoul(Timestamp regDate){
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        sdf.setTimeZone(TimeZone.getTimeZone("Asia/Seoul")); // UTC+9 타임존 설정

        return sdf.format(regDate);
    }

}

File: frontend/simda/lib/models/FeedDto.dart
import 'package:simda/models/CommentDto.dart';

import 'UserDto.dart';

class FeedDto {
  final String content;
  final int emotion;
  final int feedId;
  final String img;
  final double lat;
  late int likeCnt;
  final double lng;
  final String nickname;
  final String regDate;
  final String title;
  final int userId;
  late List<CommentDto>? comments = [];

  FeedDto({
    required this.content,
    required this.emotion,
    required this.feedId,
    required this.img,
    required this.lat,
    required this.likeCnt,
    required this.lng,
    required this.nickname,
    required this.regDate,
    required this.title,
    required this.userId,
    this.comments
  });

  factory FeedDto.fromJson(Map<String, dynamic> json) {
    List<CommentDto> comments = [];
    if(json['comments'] != null){
      print('comments not null');
      json['comments'].forEach((commentJson){
        comments.add(CommentDto.fromJson(commentJson));
      });
    }
    print('comments null');

    return FeedDto(
      content: json['content'],
      emotion: json['emotion'],
      feedId: json['feedId'],
      img: json['img'],
      lat: json['lat'],
      likeCnt: json['likeCnt'],
      lng: json['lng'],
      nickname: json['nickname']??'',
      regDate: json['regDate']??'',
      title: json['title'],
      userId: json['userId'],
      comments: comments,
    );}

  Map<String, dynamic> toJson() {
    return {
      "content": content,
      "emotion": emotion,
      "feedId": feedId,
      "img": img,
      "lat": lat,
      "likeCnt": likeCnt,
      "lng": lng,
      "nickname": nickname,
      "regDate": regDate,
      "title": title,
      "userId": userId,
      "comments" : comments!.map((comment) => comment.toJson()).toList(),
    };
  }
}